File: src/test/java/uk/gov/justice/hmpps/casenotes/services/CaseNoteTypeMergerTest.java
package uk.gov.justice.hmpps.casenotes.services;

import org.junit.Test;
import uk.gov.justice.hmpps.casenotes.dto.CaseNoteType;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class CaseNoteTypeMergerTest {

    private final CaseNoteTypeMerger merger = new CaseNoteTypeMerger();

    @Test
    public void testSimpleMerge() {

        final var list1 = List.of(
                CaseNoteType.builder().code("OBS").description("Observation").subCodes(
                        List.of(CaseNoteType.builder().code("GEN").description("General").build(), CaseNoteType.builder().code("SP").description("Special").build())
                ).build(),
                CaseNoteType.builder().code("KA").description("Key worker").subCodes(
                        List.of(CaseNoteType.builder().code("KS").description("Session").build(), CaseNoteType.builder().code("KE").description("Entry").build())
                ).build()
        );

        final var list2 = List.of(
                CaseNoteType.builder().code("OBS").description("Observation").subCodes(
                        List.of(CaseNoteType.builder().code("NEW").description("New Stuff").build(), CaseNoteType.builder().code("GEN").description("Different Gen").build())
                ).build(),
                CaseNoteType.builder().code("POM").description("POM Stuff").subCodes(
                        List.of(CaseNoteType.builder().code("SPC").description("Special").build(), CaseNoteType.builder().code("GEN").description("General").build())
                ).build()
        );

        List<CaseNoteType> resultantList = merger.mergeAndSortList(list1, list2);

        assertThat(resultantList).hasSize(3);

        assertThat(resultantList).containsExactly(
                List.of(
                        CaseNoteType.builder().code("KA").description("Key worker").subCodes(
                                List.of(CaseNoteType.builder().code("KE").description("Entry").build(), CaseNoteType.builder().code("KS").description("Session").build())
                        ).build(),
                        CaseNoteType.builder().code("OBS").description("Observation").subCodes(
                                List.of(CaseNoteType.builder().code("GEN").description("Different Gen").build(), CaseNoteType.builder().code("NEW").description("New Stuff").build(), CaseNoteType.builder().code("SP").description("Special").build())
                        ).build(),
                        CaseNoteType.builder().code("POM").description("POM Stuff").subCodes(
                                List.of(CaseNoteType.builder().code("GEN").description("General").build(), CaseNoteType.builder().code("SPC").description("Special").build())
                        ).build()
                ).toArray(new CaseNoteType[0])
        );
    }

    @Test
    public void testActiveFlagUpdatedWhenParentInactiveDuringMerge() {

        final var list1 = List.of(
                CaseNoteType.builder().code("OBS").description("Observation").activeFlag("Y").subCodes(
                        List.of(CaseNoteType.builder().code("GEN").description("General").activeFlag("Y").build(), CaseNoteType.builder().code("SP").description("Special").activeFlag("N").build())
                ).build(),
                CaseNoteType.builder().code("DRR").description("Drug Rehabilitation Requirement").activeFlag("N").subCodes(
                        List.of(CaseNoteType.builder().code("DCOUN").description("Drug Counselling Session").activeFlag("Y").build(), CaseNoteType.builder().code("DTEST").description("Drug Test").activeFlag("N").build())
                ).build()
        );

        final var list2 = List.of(
                CaseNoteType.builder().code("OBS").description("Observation").activeFlag("Y").subCodes(
                        List.of(CaseNoteType.builder().code("NEW").description("New Stuff").activeFlag("Y").build(), CaseNoteType.builder().code("GEN").activeFlag("Y").description("Different Gen").activeFlag("Y").build())
                ).build(),
                CaseNoteType.builder().code("POM").description("POM Stuff").activeFlag("Y").subCodes(
                        List.of(CaseNoteType.builder().code("SPC").description("Special").activeFlag("Y").build(), CaseNoteType.builder().code("GEN").description("General").activeFlag("Y").build())
                ).build()
        );

        List<CaseNoteType> resultantList = merger.mergeAndSortList(list1, list2);

        assertThat(resultantList).hasSize(3);

        assertThat(resultantList).containsExactly(
                List.of(
                        CaseNoteType.builder().code("DRR").description("Drug Rehabilitation Requirement").activeFlag("N").subCodes(
                                List.of(CaseNoteType.builder().code("DCOUN").description("Drug Counselling Session").activeFlag("N").build(), CaseNoteType.builder().code("DTEST").description("Drug Test").activeFlag("N").build())
                        ).build(),
                        CaseNoteType.builder().code("OBS").description("Observation").activeFlag("Y").subCodes(
                                List.of(CaseNoteType.builder().code("GEN").description("Different Gen").activeFlag("Y").build(), CaseNoteType.builder().code("NEW").description("New Stuff").activeFlag("Y").build(), CaseNoteType.builder().code("SP").description("Special").activeFlag("N").build())
                        ).build(),
                        CaseNoteType.builder().code("POM").description("POM Stuff").activeFlag("Y").subCodes(
                                List.of(CaseNoteType.builder().code("GEN").description("General").activeFlag("Y").build(), CaseNoteType.builder().code("SPC").description("Special").activeFlag("Y").build())
                        ).build()
                ).toArray(new CaseNoteType[0])
        );
    }
}

File: src/test/java/uk/gov/justice/hmpps/casenotes/services/CaseNoteTypeMergerTest.kt
package uk.gov.justice.hmpps.casenotes.services

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import uk.gov.justice.hmpps.casenotes.dto.CaseNoteType

class CaseNoteTypeMergerTest {
  private val merger = CaseNoteTypeMerger()

  @Test
  fun testSimpleMerge() {
    val list1 = listOf(
        createBuilder("OBS", "Observation").subCodes(
            listOf(create("GEN", "General"), create("SP", "Special"))
        ).build(),
        createBuilder("KA", "Key worker").subCodes(
            listOf(create("KS", "Session"), create("KE", "Entry"))
        ).build()
    )
    val list2 = listOf(
        createBuilder("OBS", "Observation").subCodes(
            listOf(create("NEW", "New Stuff"), create("GEN", "Different Gen"))
        ).build(),
        createBuilder("POM", "POM Stuff").subCodes(
            listOf(create("SPC", "Special"), create("GEN", "General"))
        ).build()
    )
    val resultantList = merger.mergeAndSortList(list1, list2)
    Assertions.assertThat(resultantList).containsExactly(
        createBuilder("KA", "Key worker").subCodes(
            listOf(create("KE", "Entry"), create("KS", "Session"))
        ).build(),
        createBuilder("OBS", "Observation").subCodes(
            listOf(create("GEN", "Different Gen"), create("NEW", "New Stuff"), create("SP", "Special"))
        ).build(),
        createBuilder("POM", "POM Stuff").subCodes(
            listOf(create("GEN", "General"), create("SPC", "Special"))
        ).build()
    )
  }

  @Test
  fun testActiveFlagUpdatedWhenParentInactiveDuringMerge() {
    val list1 = listOf(
        createBuilder("OBS", "Observation").subCodes(
            listOf(create("GEN", "General"), createInactive("SP", "Special"))
        ).build(),
        createBuilder("DRR", "Drug Rehabilitation Requirement").activeFlag("N").subCodes(
            listOf(create("DCOUN", "Drug Counselling Session"), createInactive("DTEST", "Drug Test"))
        ).build()
    )
    val list2 = listOf(
        createBuilder("OBS", "Observation").subCodes(
            listOf(create("NEW", "New Stuff"), create("GEN", "Different Gen"))
        ).build(),
        createBuilder("POM", "POM Stuff").subCodes(
            listOf(create("SPC", "Special"), create("GEN", "General"))
        ).build(),
        createBuilder("DRR", "Drug Rehabilitation Requirement").subCodes(
            listOf(create("DCOUN", "Drug Counselling Session"))
        ).build()
    )
    val resultantList = merger.mergeAndSortList(list1, list2)
    Assertions.assertThat(resultantList).containsExactly(
        createBuilder("DRR", "Drug Rehabilitation Requirement").subCodes(
            listOf(create("DCOUN", "Drug Counselling Session"), createInactive("DTEST", "Drug Test"))
        ).build(),
        createBuilder("OBS", "Observation").subCodes(
            listOf(create("GEN", "Different Gen"), create("NEW", "New Stuff"), createInactive("SP", "Special"))
        ).build(),
        createBuilder("POM", "POM Stuff").subCodes(
            listOf(create("GEN", "General"), create("SPC", "Special"))
        ).build()
    )
  }

  private fun createBuilder(obs: String, observation: String): CaseNoteType.CaseNoteTypeBuilder {
    return CaseNoteType.builder().code(obs).description(observation)
  }

  private fun createInactive(sp: String, special: String): CaseNoteType {
    return createBuilder(sp, special).activeFlag("N").build()
  }

  private fun create(gen: String, general: String): CaseNoteType {
    return createBuilder(gen, general).build()
  }
}
